designs/2019-public-config-array-api/README.md

```diff
@@ -0,0 +1,85 @@
+- Start Date: 2019-12-09
+- RFC PR: <!-- leave this empty, to be filled in later -->
+- Authors: @manuth
+
+# Public `ConfigArray` API
+
+## Summary
+
+Allow 3rd-party developers to determine which configuration-files are considered by `eslint`.
+
+## Motivation
+
+In some cases people might want to automatically trigger a new linting process as soon as the configuration of eslint has changed.
+This is a key feature for language servers which provide warnings and fixes to IDEs.
+
+Therefore the language-server needs to know which config-files are considered by `eslint`.
+
+## Detailed Design
+
+A method called `getConfigArrayForFile()` shall be added to the `CLIEngine` class.
+By providing such a feature people can easily watch the correct files for changes as soon as the configuration changes.
+
+***Example:***
+```js
+import fs = require("fs");
+
+let engine = new CLIEngine({});
+let file = "./App/src/index.ts";
+let configArray = engine.getConfigArrayForFile(file);
+
+for (let config of configArray) {
+    if (config.filePath.length > 0) {
+        fs.watchFile(
+            conig.filePath,
+            () =>
+            {
+                lint();
+            });
+    }
+}
+
+function lint() {
+    engine.executeOnFiles([file]);
+}
+
+lint();
+```
+
+## Documentation
+
+I think it might be a good idea to blog about this new feature as this is a key feature for writing language servers for IDEs.
+That way the integration of `eslint` in all sorts of text-editors can be improved to its best.
+
+## Drawbacks
+
+The only downside I can think of is that on every change to the `ConfigArray` or `CascadingConfigArray` API either a new major version has to be released or a separate implementation of said classes has to be added for the public API.
+
+## Backwards Compatibility Analysis
+
+Adding this feature provides full backwards compatibility.
+
+## Alternatives
+
+Add a method which only returns filepaths and a value indicating whether said filepath is `root`.  
+That way the drawbacks could be fixed but the developers don't have the benefit to further analyze the config-array.
+
+A similar method already exists: The `CLIEngine#getConfigForFile`-method. Though this method does not indicate which files are considered by `eslint` which draws developers of language-servers unable to hot-reload `eslint` as soon as a config-file changes.
+
+## Open Questions
+
+  - Is it reasonable to return the whole config-array or should such a method only return the `filePath`s?
+  - Does it make sense to also return the `root`-directory in order to let language-servers run `eslint` with `root` as its `cwd`?
```

Comment:
I'd like to clarify the mean of the `root`.

1. The directory that `node_modules/eslint` exists. (I guess that API user knows it already)
1. The directory that `.eslintignore` exists (because ESLint ignores `.eslintignore` of subdirectories). (this is CWD)
1. The directory that the last config file exists. ("last" means that it's in the most parent directory of all config files because ESLint traverses directories from leaf to root to find config files.) (I'm not sure if this path is meaningful for something.)
1. Others.

This question may be related to https://github.com/eslint/rfcs/pull/47.

Reply:
So in case of the language server functionality the `root` definitely should be returned.
Shall we create a separate issue for the language server stuff or just rename the current RFC properly and handle both, the public config-array API and the API for language servers?

Reply:
I guess that detecting the root is the responsibility of language servers rather than eslint. ESLint doesn't have such a `root` concept.

The `root` flag of config files is just a command to stop traversing of config file detection. It's not meaning the root directory of projects. For example, even if there is the `root` flag in a subdirectory, ESLint still load `.eslintignore` and plugins from CWD. The CWD is similar to the root and it's given from outside of ESLint.

Reply:
Oh well that's fair.
I think, in this case I will just go for the workspace-folder by default as anyone would usually run `eslint` in the workspace-folder.

Thanks for your advice!